### ISSUE TITLE
Globally limit concurrent access to files in the repo

### ISSUE THREAD

**Issue by cd1188** [restic/restic#1763]
## Output of `restic version`
restic 0.8.3 (v0.8.3-221-g26757ae) compiled with go1.10.1 on windows/amd64

## How did you run restic exactly?
`restic -r i:\backup check --read-data`

## What backend/server/service did you use to store the repository?

external HDD-drive with usb3

## Expected behavior

reading/checking the data with full speed of HDD

## Actual behavior

backup need 2h of time
checking with --read-data option took 15h!

## Steps to reproduce the behavior
`restic -r i:\backup check --read-data`

## Do you have any idea what may have caused this?
if restic run, the windows-task-manager shows me a hdd-workload of 100% and reading only with 7MB/s

i found in the file `internal/checker.go` the variable `defaultParallelism = 40`
so that means 40 thread for reading on the disk.. right?

accessing 40 files at the same time is for a HDD too much workload.

## Do you have an idea how to solve the issue?

if i set defaultParallelism=1 and compile it, restic need then much more time for starting:
'load indexes'
'check all packs'
'check snapshots, trees and blobs'

but after that.. reading/checking data ist much faster.
the windows-task-manager shows me a hdd-workload of 50% and reading speed is 45MB/s

with defaultParallelism=2
the windows-task-manager shows me a hdd-workload at 80-90% and
reading speed is 35-50MB/s

so i think to protect the hdd-drive i would prefer to set defaultParallelism=1 in that case.
my solution for faster start is to leave defaultParallelism=40   (or maybe 4 ?)
and add a defaultParallelism2=1 only for the ReadPacks-Function in the that file.

you musst me sorry for this description, beacause i am not a GIT-user and also not good in english
i changed for me this code and its working very well with a speed-up of factor 6
maybe you can change this in the master-branch?

EDIT:
after testing it on an SSD .. the cpu is now the bottleneck and here would be more worker way faster.
maybe a command-switch would be better like --read-worker 4 or --ssd
-> in this case more worker then cpu-threads are useless for hashing/checking ??
but for regular-hdd i would stay by 1 because the health of the hdd is more important fore me.

	// run workers
	const defaultParallelism2 = 1                    // NEW
	for i := 0; i < defaultParallelism2; i++ {     // CHANGED
		g.Go(func() error {
			for {
				var id restic.ID
				var ok bool

				select {
				case <-ctx.Done():
					return nil
				case id, ok = <-ch:
					if !ok {
						return nil
					}
				}

				err := checkPack(ctx, c.repo, id)
				p.Report(restic.Stat{Blobs: 1})
				if err == nil {
					continue
				}

				select {
				case <-ctx.Done():
					return nil
				case errChan <- err:
				}
			}
		})
	}

## Did restic help you or made you happy in any way?

restic is superb! :-)

**Comment by fd0** [comment]
Thanks for the report, that's something we need to address.

**Comment by bjoe2k4** [comment]
Sorry to poke on this, but is there any update? This seems to be quite a show-stopper for any HDD storage user, as i just ran into this issue myself. According to #322, this should be tackled when the new archiver is implemented, which is the case :-)

**Comment by alphapapa** [comment]
It would be nice to be able to limit the number of threads.  e.g. when my desktop system is busy and a scheduled `restic check` runs, even though I have restic running with `nice -n 19`, it still reduces the system's interactive responsiveness, because every core is saturated.

**Comment by fd0** [comment]
@alphapapa while this is not a solution per se, did you try running restic with both `nice` and `ionice`?

**Comment by alphapapa** [comment]
Yes, I use `nice -n 19 ionice -c3` on many background jobs, including Restic.  The problem in this case is that the CPU is already at about 50% usage, and when the `restic check` job runs, the niced threads saturate the rest of the CPU.  The Linux scheduler tries its best, but the non-niced processes still suffer reduced performance.  In this case, the CPU was already busy playing video in a browser (which takes too much CPU compared to playing it directly in VLC, but I digress), and when Restic started, the video playback became very choppy until Restic terminated.

**Comment by fd0** [comment]
@alphapapa then this is the wrong issue: it is about limiting the number of concurrent read operations, especially on harddisks. You're talking about limiting the number of concurrent operations which are heavy on the CPU (mainly hashing and decrypting stuff).

**Comment by alphapapa** [comment]
I'd like to be able to control both, actually.  I guess I'll go look for the appropriate issue to repeat my comment, then.  :)  ...

**Comment by alphapapa** [comment]
This issue seems either a duplicate of or closely related to https://github.com/restic/restic/issues/436.  Also https://github.com/restic/restic/issues/322.

@fd0 Other than those, I don't see an issue specifically about limiting the number of threads used by `restic check`.  Should I post a new issue requesting this feature?

**Comment by fd0** [comment]
Hm, I think we need to address limiting concurrent access to resources (disc, CPU, network) globally. Let's track this in this issue for now.

**Comment by stbisplinghoff** [comment]
I ran into this issue today when I replaced my plain old USB2 HDDs with larger and newer USB3 HDDs. Before I achieved around 20MB/s (on Windows) using check with read-data and I am surprised now, that this value does not change at all. 

I am aware that for v0.9.6 the defaultParallelism value has been reduced from 40 to 5, however my HDD still makes lot of noise indicating the head is jumping around. Disc access monitoring also shows me that pack files are selected randomly. In comparison, using 7zip hash calculation yields >100MB/s when sequentially reading the pack files.

Although I would be happier if check read-data could achieve the same read performance, would it be ok to use 7zip for checking pack integrity against data corruption? Of course, this requires that the pack storage has been proven sane before, can this be achieved faster than running check read-data on the whole repository?

**Comment by senotrusov** [comment]
Just to add my personal data here.

With the current ``defaultParallelism = 5`` the data check on my USB3 HDD performs with the speed of ~6 MB/sec. For 1TB of data that's about two days of work, and an extra stress on the HDD. I changed ``defaultParallelism = 1`` and compiled my own restic (which is quite easy to do), and I now I got speed around 35 MB/sec.

I understand that ``defaultParallelism = 5`` is the preferred setting for the remote repositories. I saw some discussions about setting that preference automatically. I think it will be much better to have a manual setting for that.

I fear that it might be a lot of edge cases for the automated detection. Say you have a fast LAN, and you have a spinning rust HDD which you access through that network. So it's remote in the sense that it is not on machine the restic runs on. But you might still prefer to have ``defaultParallelism = 1`` in that cause, because your LAN latency is much less than your storage random I/O latency.

Thank you for your time.

**Comment by boison-ranga** [comment]
10 parallel threads make lower-performance ARM NAS models unusable. The payload data had a size of 150GB, which already took 15 hours without having finished yet, so I needed to abort the job.

I'll now try again with `nice -n19` priority (`ionice` is not available on the model), but I suppose it won't work that better. If not successful, I probably have to switch to borgbackup, as it works strictly sequentially. 

In summary: Lower-performance model options like adjusting `defaultParallelism` would be highly appreciated. restic is awesome, I would love to stay!

**Comment by MichaelEischer** [comment]
I've opened PR #3475 which allows limiting the number of concurrent operations for the local and sftp backend. To limit the number of CPUs cores used by restic, just set the environment variable `GOMAXPROCS` accordingly.

**Comment by lrvl** [comment]
Would automatic cool down to wait for the five minute Loadavg to drop below one work in this case? I used this method quite successfully while executing massive data migration batch jobs using various tools like rclone. Before handing off the next job just wait for the medium term load to recover to normal levels.

**Comment by clst** [comment]
> just set the environment variable `GOMAXPROCS` accordingly.

Hi there. This does not seem to influence the building of an index during `GOMAXPROCS=1 restic prune`. Only the counting files stage was running single threaded. 10 Threads on a Raspberry with USB HDD is very painful. I hope the concurrency PR is ready soon.
Is it considered stable enough for production already? Then I would consider building it myself.

Thanks for the hard work.
